File: app/src/main/java/org/thoughtcrime/securesms/components/AccessibleToggleButton.java
package org.thoughtcrime.securesms.components;


import android.content.Context;
import android.util.AttributeSet;

import androidx.appcompat.widget.AppCompatToggleButton;

public class AccessibleToggleButton extends AppCompatToggleButton {

  private OnCheckedChangeListener listener;

  public AccessibleToggleButton(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
  }

  public AccessibleToggleButton(Context context, AttributeSet attrs) {
    super(context, attrs);
  }

  public AccessibleToggleButton(Context context) {
    super(context);
  }

  @Override
  public void setOnCheckedChangeListener(OnCheckedChangeListener listener) {
    super.setOnCheckedChangeListener(listener);
    this.listener = listener;
  }

  public void setChecked(boolean checked, boolean notifyListener) {
    if (!notifyListener) {
      super.setOnCheckedChangeListener(null);
    }

    super.setChecked(checked);

    if (!notifyListener) {
      super.setOnCheckedChangeListener(listener);
    }
  }

  public OnCheckedChangeListener getOnCheckedChangeListener() {
    return this.listener;
  }

}

File: app/src/main/java/org/thoughtcrime/securesms/components/AccessibleToggleButton.kt
package org.thoughtcrime.securesms.components

import android.content.Context
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatToggleButton

class AccessibleToggleButton @JvmOverloads constructor(
  context: Context,
  attrs: AttributeSet? = null,
  defStyleAttr: Int = 0
) : AppCompatToggleButton(context, attrs, defStyleAttr) {

  private var listener: OnCheckedChangeListener? = null

  override fun setOnCheckedChangeListener(listener: OnCheckedChangeListener?) {
    super.setOnCheckedChangeListener(listener)
    this.listener = listener
  }

  fun setChecked(checked: Boolean, notifyListener: Boolean) {
    if (!notifyListener) {
      super.setOnCheckedChangeListener(null)
    }

    super.setChecked(checked)

    if (!notifyListener) {
      super.setOnCheckedChangeListener(listener)
    }
  }

  fun getOnCheckedChangeListener(): OnCheckedChangeListener? {
    return this.listener
  }
}
